src/ngx_http_lua_pipe.c

```diff
@@ -0,0 +1,2476 @@
+            break;
+        }
+
+        if (n == 0) {
+            pipe_ctx->eof = 1;
+            ngx_log_debug1(NGX_LOG_DEBUG_HTTP, ngx_cycle->log, 0,
+                           "lua pipe closed pipe:%p", pipe_ctx);
+            continue;
+        }
+
+        if (n == NGX_ERROR) {
+            ngx_log_debug1(NGX_LOG_DEBUG_HTTP, ngx_cycle->log, ngx_errno,
+                           "lua pipe read data error pipe:%p", pipe_ctx);
+
+            pipe_ctx->err_type = PIPE_ERR_SYSCALL;
+            pipe_ctx->pipe_errno = ngx_errno;
+            return NGX_ERROR;
+        }
+
+        b->last += n;
+    }
+
+    return NGX_AGAIN;
+}
+
+
+static ngx_int_t
+ngx_http_lua_pipe_init_ctx(ngx_http_lua_pipe_ctx_t **pipe_ctx_pt, int fd,
+    ngx_pool_t *pool, u_char *errbuf, size_t *errbuf_size)
+{
+    ngx_connection_t                   *c;
+
+    *pipe_ctx_pt = ngx_pcalloc(pool, sizeof(ngx_http_lua_pipe_ctx_t));
+    if (*pipe_ctx_pt == NULL) {
+        *errbuf_size = ngx_snprintf(errbuf, *errbuf_size, "no memory") - errbuf;
+        return NGX_ERROR;
+    }
+
+    if (fd == -1) {
+        /* consider as closed */
+        return NGX_OK;
+    }
+
+    c = ngx_get_connection(fd, ngx_cycle->log);
+    if (c == NULL) {
+        *errbuf_size = ngx_snprintf(errbuf, *errbuf_size, "no connection")
+                       - errbuf;
+        return NGX_ERROR;
+    }
+
+    c->log = ngx_cycle->log;
+    c->recv = ngx_http_lua_pipe_fd_read;
+    c->read->handler = ngx_http_lua_pipe_dummy_event_handler;
+    c->read->log = c->log;
+
+#ifdef HAVE_SOCKET_CLOEXEC_PATCH
+    c->read->skip_socket_leak_check = 1;
+#endif
+
+    c->send = ngx_http_lua_pipe_fd_write;
+    c->write->handler = ngx_http_lua_pipe_dummy_event_handler;
+    c->write->log = c->log;
+    (*pipe_ctx_pt)->c = c;
+
+    ngx_log_debug2(NGX_LOG_DEBUG_HTTP, ngx_cycle->log, 0,
+                   "lua pipe init pipe ctx:%p fd:*%d", *pipe_ctx_pt, fd);
+
+    return NGX_OK;
+}
+
+
+int
+ngx_http_lua_ffi_pipe_proc_read(ngx_http_request_t *r,
+    ngx_http_lua_ffi_pipe_proc_t *proc, int from_stderr, int reader_type,
+    size_t length, u_char **buf, size_t *buf_size, u_char *errbuf,
+    size_t *errbuf_size)
+{
+    int                                 rc;
+    ngx_msec_t                          timeout;
+    ngx_event_t                        *rev;
+    ngx_connection_t                   *c;
+    ngx_http_lua_ctx_t                 *ctx;
+    ngx_http_lua_pipe_t                *pipe;
+    ngx_http_lua_co_ctx_t              *wait_co_ctx;
+    ngx_http_lua_pipe_ctx_t            *pipe_ctx;
+
+    rc = ngx_http_lua_pipe_get_lua_ctx(r, &ctx, errbuf, errbuf_size);
+    if (rc != NGX_OK) {
+        return rc;
+    }
+
+    ngx_log_debug2(NGX_LOG_DEBUG_HTTP, r->connection->log, 0,
+                   "lua pipe read process:%p pid:%P", proc, proc->_pid);
+
+    pipe = proc->pipe;
+    if (pipe == NULL || pipe->closed) {
+        *errbuf_size = ngx_snprintf(errbuf, *errbuf_size, "closed") - errbuf;
```

Comment:
I could not really agree that distinguishing `exited` from `closed` is a good idea...
In caller's perspective, there is not much difference between `closed` and `exited`, and checking `err ~= 'closed` and err ~= 'exited'`  is cumbersome.

@agentzh, what is your idea?

Reply:
> In caller's perspective, there is not much difference between closed and exited

I think I need to clarify my words...
Whether the process is exited or not doesn't affect to the read/write operations. The caller is still able to read/write the pipe buffer even after the target process is exited. (You will get a `EPIPE` when writing to such buffer, but writing to a closed stdin also gets the same error) The reason why I chose `pipe->closed` and returned `closed` is that the `pipe` is actually closed by the caller manually.

Reply:
> In caller's perspective, there is not much difference between closed and exited, and checking err ~= 'closed‘ and err ~= 'exited' is cumbersome.

Yes, that is exactly what I meant, and I wanted to avoid the some thus APIs had to check for `"exited"` and others for `"closed"` (which could easily lead to user errors...).

That said, I did realize that each error message had a precise meaning:

> There are some instances when we return "closed" in which it is fine, of course (e.g. a pipe is closed, which is different from the process being exited).

However, in hindsight to what was said in this thread and the other one, I do agree with this design. The documentation can be clarified to explain this, I will make sure to do so when reviewing the docs in particular (I have yet to push my changes and comments there for now, but it is ongoing).